Fixed: 12012: Add Support for `TABLESPACE` Specification in Table Definitions for Oracle

Ineffable22 opened the pull request:
Fixes: #12016
**Description**

This PR adds support for specifying the `TABLESPACE` in table definitions in SQLAlchemy, specifically for Oracle. This feature is particularly useful for Oracle users who need to specify the tablespace where the table data will be stored.

**Changes Made**
1. Updated `construct_arguments` in `OracleDialect`:
    - The `construct_arguments` list in the `OracleDialect` class has been updated to include the `tablespace` argument for the `Table` class.

```Python
construct_arguments = [
    (
        sa_schema.Table,
        # old
        {"resolve_synonyms": False, "on_commit": None, "compress": False},
        # new
        {"resolve_synonyms": False, "on_commit": None, "compress": False, "tablespace": None},
    ),
    (sa_schema.Index, {"bitmap": False, "compress": False}),
]
```
**Path**: `lib/sqlalchemy/dialects/oracle/base.py`

2. Modified OracleDDLCompiler to Include TABLESPACE in post_create_table:
    - The OracleDDLCompiler class has been modified to include the TABLESPACE clause at the end of the CREATE TABLE statement if the tablespace option is provided.

```Python
if opts["tablespace"]:
    tablespace_name = opts["tablespace"]
    table_opts.append(
        "\n TABLESPACE %s" % self.preparer.quote(tablespace_name)
    )
```
**Path**: `lib/sqlalchemy/dialects/oracle/base.py`

3. Added tablespace Argument to the Table Class:
    - A new tablespace argument has been added to the Table class to allow specifying the tablespace in the table definition.

4. Documentation Update:
    - The documentation has been updated to reflect the new feature and provide usage examples.

**Usage Example**

```Python
from sqlalchemy import create_engine, Table, Column, Integer, String, MetaData

engine = create_engine('oracle+cx_oracle://user:password@dsn')

metadata = MetaData()

users = Table('users', metadata,
    Column('id', Integer, primary_key=True),
    Column('name', String),
    Column('email', String, unique=True),
    oracle_tablespace='my_tablespace'  # New tablespace argument optional
)

metadata.create_all(engine)
```

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

Ineffable22 committed:
Pull Request: Add Support for `TABLESPACE` Specification in Table Definitions for Oracle

**Description**

This PR adds support for specifying the `TABLESPACE` in table definitions in SQLAlchemy, specifically for Oracle. This feature is particularly useful for Oracle users who need to specify the tablespace where the table data will be stored.

**Changes Made**
1. Updated `construct_arguments` in `OracleDialect`:
    - The `construct_arguments` list in the `OracleDialect` class has been updated to include the `tablespace` argument for the `Table` class.

```Python
construct_arguments = [
    (
        sa_schema.Table,
        # old
        {"resolve_synonyms": False, "on_commit": None, "compress": False},
        # new
        {"resolve_synonyms": False, "on_commit": None, "compress": False, "tablespace": None},
    ),
    (sa_schema.Index, {"bitmap": False, "compress": False}),
]
```
**Path**: `lib/sqlalchemy/dialects/oracle/base.py`

2. Modified OracleDDLCompiler to Include TABLESPACE in post_create_table:
    - The OracleDDLCompiler class has been modified to include the TABLESPACE clause at the end of the CREATE TABLE statement if the tablespace option is provided.

```Python
if opts["tablespace"]:
    tablespace_name = opts["tablespace"]
    table_opts.append(
        "\n TABLESPACE %s" % self.preparer.quote(tablespace_name)
    )
```
**Path**: `lib/sqlalchemy/dialects/oracle/base.py`

3. Added tablespace Argument to the Table Class:
    - A new tablespace argument has been added to the Table class to allow specifying the tablespace in the table definition.

4. Documentation Update:
    - The documentation has been updated to reflect the new feature and provide usage examples.

**Usage Example**

```Python
from sqlalchemy import create_engine, Table, Column, Integer, String, MetaData

engine = create_engine('oracle+cx_oracle://user:password@dsn')

metadata = MetaData()

users = Table('users', metadata,
    Column('id', Integer, primary_key=True),
    Column('name', String),
    Column('email', String, unique=True),
    oracle_tablespace='my_tablespace'  # New tablespace argument optional
)

metadata.create_all(engine)
```

CaselIT commented:
Thanks for the PR.

Could you also
- add some tests, like these https://github.com/sqlalchemy/sqlalchemy/blob/7495b51f9ae53b3a5c631d38397f5b9fece52015/test/dialect/oracle/test_compiler.py#L1470
- document it here? https://docs.sqlalchemy.org/en/20/dialects/oracle.html#oracle-table-options (that section should be in the base.py module doc string)

Ineffable22 committed:
Add tests for Oracle tablespace specification in table definitions

This commit adds new tests to verify the correct handling of tablespace names in Oracle table definitions. The tests ensure that tablespace names are recognized correctly regardless of their case (uppercase, lowercase, mixed case).

Tests added:
- : Verifies that tablespace names are correctly included in the CREATE TABLE statement for different cases:
  - Uppercase tablespace name
  - Lowercase tablespace name
  - Mixed case tablespace name

These tests ensure that the  argument is handled correctly and that the generated SQL statements include the specified tablespace names as expected.

CaselIT commented:
Thanks for adding the test.
If you can mention this new option here https://github.com/sqlalchemy/sqlalchemy/blob/aaddd7c8403e9ca2f77113467b5e2ae279a542c4/lib/sqlalchemy/dialects/oracle/base.py#L479-L501 so that it is visible in the docs.

Also it would be nice if you could add a changelog here (the number is the issue, so 12016) https://github.com/sqlalchemy/sqlalchemy/tree/main/doc/build/changelog/unreleased_20

You find an example here already. Something like

> Added new table option ``oracle_tablespace`` to specify the ``TABLESPACE`` option of an
table in Oracle. Pull request courtesy of Miguel Grillo.

Ineffable22 committed:
docs: add documentation for oracle_tablespace option

Added documentation for the new table option  to specify the
 option when creating a table in Oracle. This includes an example
usage and a detailed description of the feature. Pull request courtesy of Miguel Grillo.

For more details, see the discussion at https://github.com/sqlalchemy/sqlalchemy/issues/12016.

Ineffable22 commented:
Hi,

I have added the documentation for the `oracle_tablespace` option in the `base.py` . Additionally, I have updated the changelog as requested.

Thank you!

CaselIT commented:
that's great, thanks! I'll make the CI run etc.

Thank you for your effort!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 466295b2b41d229827f365500569ff019eca2f27 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 466295b2b41d229827f365500569ff019eca2f27: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

I'll cherry pick to 2_0 once the ci passes

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

Ineffable22 committed:
style: fix E501 line too long in base.py. Refactored a line in base.py to comply with PEP 8 E501 (line too long) by breaking it into multiple lines.

Ineffable22 commented:
Hi,

I have refactored a line in `base.py` to comply with PEP 8 E501 (line too long) by breaking it into multiple lines.

Thank you!

CaselIT commented:
it seems there are still format issues. Can you run black on the files you changed?

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

* lib/sqlalchemy/dialects/oracle/base.py (line 497): note that we currently reflect oracle_compress, but not on_commit.  so the latter gives us some precedent to not reflect tablespace either.  but we could

CaselIT commented:
> note that we currently reflect oracle_compress, but not on_commit. so the latter gives us some precedent to not reflect tablespace either. but we could

seems easy do reflect. will add to the PR

Ineffable22 committed:
style: fix formatting issue in base.py

Added a missing comma to ensure the code passes black formatting.

Ineffable22 committed:
style: reformat test_table_tablespace

Ineffable22 commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -500,6 +500,15 @@
+    The ``oracle_tablespace`` parameter specifies the tablespace in which the
+    table is to be created. This is useful when you want to create a table in a
+    tablespace other than the default tablespace of the user.
+
```
Hi,

Is .. versionadded:: 2.0.36 correct for this change?

Thank you!

CaselIT replied:
2.0.37 since it's the next one

Ineffable22 replied:
Great, added the .. versionadded:: 2.0.37

sqla-tester replied:
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

Ineffable22 committed:
test: refactor test_table_tablespace to use testing.combinations

Refactored the test_table_tablespace to use testing.combinations for better maintainability and extensibility. This change reduces code repetition and makes it easier to add new test cases in the future.

Ineffable22 commented on test/dialect/oracle/test_compiler.py:
```diff
@@ -1659,6 +1659,46 @@ def test_double_to_oracle_double(self):
             cast(column("foo"), d1), "CAST(foo AS DOUBLE PRECISION)"
         )
 
+    def test_table_tablespace(self):
```
Hi, I just refactored the test

Ineffable22 committed:
docs: add versionadded directive for oracle_tablespace option

Added the .. versionadded:: 2.0.37 directive to document the version in which the
oracle_tablespace option was introduced.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e542dea6ced269cb188b06cbd04cecf1c400e29a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset e542dea6ced269cb188b06cbd04cecf1c400e29a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

CaselIT commented:
thanks @Ineffable22

Ineffable22 commented:
@CaselIT You're welcome! Happy to contribute.

Ineffable22 commented:
Hi,

When will version 2.0.37 be available with these changes or do you need me to add or update anything else?

Thanks

sqla-tester commented on test/dialect/oracle/test_compiler.py:
```diff
@@ -1659,6 +1659,26 @@ def test_double_to_oracle_double(self):
             cast(column("foo"), d1), "CAST(foo AS DOUBLE PRECISION)"
         )
 
+    @testing.combinations(
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

* lib/sqlalchemy/dialects/oracle/base.py (line 497): Done
* lib/sqlalchemy/dialects/oracle/base.py (line 2203): The side effect of this is that now most tables include this value.

Oracle has the concept of "default tablespace" but:
- it seems to be configurable in the database or per-user
- requires another query to dba_users (all users does not have that column). not sure if there is another option here.

So I think it's likely fine to have this field almost always populated

CaselIT commented:
> Hi,
> 
> When will version 2.0.37 be available with these changes or do you need me to add or update anything else?
> 
> Thanks

likely in a couple of weeks. I've added reflection support on gerrit today

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

ready for review mike

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5550

* lib/sqlalchemy/dialects/oracle/base.py (line 2203): as mentioned in the dev meeting we will merge like this and maybe do something different if we get complains

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5582 has been **merged**. Congratulations! :)
